File: 0088-merge-sorted-array/0088-merge-sorted-array.cpp
class Solution {
public:
//merge the aar
//optimize 1st swapping the elements 

    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        //store smaller elements in arr1 and larger in arr 2
        int left=m-1,right=0;
        while(left>=0 && right<n){
            if(nums1[left]>nums2[right]){
                swap(nums1[left],nums2[right]);
                left--;right++;
            }else break;
        }
        sort(nums1.begin(),nums1.begin()+m);
        sort(nums2.begin(),nums2.begin()+n);
        for(int i=0;i<n;i++){
        nums1[m+i]=nums2[i];
        }
    }
};

File: 0088-merge-sorted-array/0088-merge-sorted-array.java
class Solution {
    void swap(int[] nums1, int ind1, int[] nums2, int ind2){
        if(nums1[ind1] > nums2[ind2]){
            int temp = nums1[ind1];
            nums1[ind1] = nums2[ind2];
            nums2[ind2] = temp;
        }
    }

    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int len = m + n;
        int gap = (len / 2) + (len % 2);

        while (gap > 0) {
            int left = 0;
            int right = gap;

            while (right < len) {
                // both in nums1
                if (left < m && right < m) {
                    if (nums1[left] > nums1[right]) {
                        int temp = nums1[left];
                        nums1[left] = nums1[right];
                        nums1[right] = temp;
                    }
                }
                // left in nums1, right in nums2
                else if (left < m && right >= m) {
                    if (nums1[left] > nums2[right - m]) {
                        int temp = nums1[left];
                        nums1[left] = nums2[right - m];
                        nums2[right - m] = temp;
                    }
                }
                // both in nums2
                else {
                    if (nums2[left - m] > nums2[right - m]) {
                        int temp = nums2[left - m];
                        nums2[left - m] = nums2[right - m];
                        nums2[right - m] = temp;
                    }
                }

                left++;
                right++;
            }

            if (gap == 1) break;
            gap = (gap / 2) + (gap % 2);
        }

        for (int i = 0; i < n; i++) {
            nums1[m + i] = nums2[i];
        }
    }
}
